core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -1004,24 +1004,28 @@ private Builder setInitialFormatVersion(int newFormatVersion) {
       return this;
     }
 
-    public Builder upgradeFormatVersion(int newFormatVersion) {
+    public Builder upgradeFormatVersion(int targetFormatVersion) {
       Preconditions.checkArgument(
-          newFormatVersion <= SUPPORTED_TABLE_FORMAT_VERSION,
+          targetFormatVersion <= SUPPORTED_TABLE_FORMAT_VERSION,
           "Cannot upgrade table to unsupported format version: v%s (supported: v%s)",
-          newFormatVersion,
+          targetFormatVersion,
           SUPPORTED_TABLE_FORMAT_VERSION);
       Preconditions.checkArgument(
-          newFormatVersion >= formatVersion,
+          targetFormatVersion >= formatVersion,
           "Cannot downgrade v%s table to v%s",
           formatVersion,
-          newFormatVersion);
+          targetFormatVersion);
 
-      if (newFormatVersion == formatVersion) {
+      if (targetFormatVersion == formatVersion) {
         return this;
       }
 
-      this.formatVersion = newFormatVersion;
-      changes.add(new MetadataUpdate.UpgradeFormatVersion(newFormatVersion));
+      // register incremental version changes separately to ensure all upgrade requirements are met
+      for (int version = formatVersion + 1; version <= targetFormatVersion; version++) {
+        changes.add(new MetadataUpdate.UpgradeFormatVersion(version));
```

Comment:
More fundamental question (which I don't think we really need to answer in this PR, but maybe just think through) cc @RussellSpitzer @leangjonathan  @rdblue @nastra :

What if a REST server gets a list of upgrades out of order, like "1->3->2". I know here in the reference implementation that's not happening because of the loop but some other implementation theoretically could send changes in that order.

Do we need to change the REST spec to require a certain failure for this case?

I think ultimately we *do not* need to require that servers have a single upgrade path.

Servers can handle the upgrade path according to what they think is best (either by re-ordering those if they want to do the upgrade iteratively or by just doing the direct upgrade path if they like). This keeps the spec small and implementations can do whatever intelligent things they like.

Reply:
+1

I was poking around the HTTP client, REST (de)serializers, and request parser logic. There is no content enforcement JSON inputs for deserialization . Specifically, the implementation of `UpdateTableRequestParser::fromJson` does not include any moderation of contents. Though I suppose deserialization and parsing wouldn't be the right place to enforce this so we come back to @nastra's concern about the REST spec and getting out of order or missing versions in the upgrade list.

Reply:
Is there currently any downside in doing an upgrade from 1 to 3 directly without having an internal upgrade to 2? I believe the only restriction we have is if you downgrade the version, but there aren't any other restrictions that would prevent us from going 1 -> 3 directly?
Going 1 -> 3 direclty means that we don't have to create multiple changes of `MetadataUpdate.UpgradeFormatVersion` and not worry about a client potentially sending such an upgrade path out-of-order

Reply:
@nastra Check out the thread https://github.com/apache/iceberg/pull/10861#discussion_r1703406710, I had a similar question earlier but the rationale for the internal stepwise upgrade is to avoid having multiple upgrade paths from a table metadata builder perspective. 

So the the rationale for performing the stepwise upgrade at least from a client side perspective in the reference implementation (the change here):

The stepwise approach has the advantage of keeping testing a bit simpler (don't need to test 1->3 direct and 1->2->3, there's only a single path from 1 to 3) . It also seems better from a TableMetadata API perspective to ensure that any potential issues on the upgrade path have a consistent failure point. So if there's an issue upgrading from 2->3 in the 1->2->3 path, it's easy to distinguish that compared to 1->3 somehow just working (as an example).


I think in the end for the server side we can't really require a single upgrade path because the client can send any list of upgrade options, nor do I think we want to spec that out since we probably want to enable servers to make their choice? If they want to do the stepwise internal upgrade themselves since they consider it more safe or if they want to direct upgrade they can just do that.

Reply:
The problem we are solving involves how we maintain different upgrade paths. Currently, this stepwise internal upgrade is for future proofing version requirements. @amogh-jahagirdar @RussellSpitzer correct me if I'm wrong.

In the logic for upgrade requirements we currently have the following pattern:
```
  public static List<UpdateRequirement> forUpdateTable(
      TableMetadata base, List<MetadataUpdate> metadataUpdates) {
    Preconditions.checkArgument(null != base, "Invalid table metadata: null");
    Preconditions.checkArgument(null != metadataUpdates, "Invalid metadata updates: null");
    Builder builder = new Builder(base, false);
    builder.require(new UpdateRequirement.AssertTableUUID(base.uuid()));
    metadataUpdates.forEach(builder::update);
    return builder.build();
  }
```
Where `metadataUpdates.forEach(builder::update);` will register UpdateRequirements based on the specific metadata update into a list for later processing.

Using the 1, 2, 3 example...

Let's say that V2 required some field XYZ to be populated and we register that as an `UpdateRequirement` subclass.
```
  class AssertXYZPopulated implements UpdateRequirement {
    ...
  }
```
And we also fill in the appropriate builder logic to link `MetadataUpdate.UpgradeFormatVersion(2)` to `AssertXYZPopulated`

For V1->V2 upgrades the original logic worked since we would have a `MetadataUpdate.UpgradeFormatVersion(2)` instance registered and can thus enforce `AssertXYZPopulated`.

However, for V1->V3 upgrades, the original logic would have only had `MetadataUpdate.UpgradeFormatVersion(3)` (which does not have the mapping to `AssertXYZPopulated`) and we would fail to enforce the V2 upgrade requirements.

By adding the stepwise logic we are ensuring future upgrade requirements are always enforced version by version.

Reply:
@leangjonathan I think the stepwise can be implemented on the upgrade itself rather than in our test framework. So if a user does Update 1 -> 3 on the back end we can do 1 -> 2 -> 3 so we necessarily would go through all the steps. I think we are ok just doing

N -> Latest and N -> N+1

The N+1 just for helping us track down stepwise pieces that are broken. Although technically if all the n -> n+1 upgrades work the n -> latest should also work if we implement it correctly.

Reply:
@nastra Do you have any other comment on this? I think we should get this and the other test stuff in asap so we can work on actual V3 Spec features

Reply:
I've reached out to @rdblue to also get his opinion on any potential issues by doing a direct 1 -> 3 upgrade or whether we indeed would need to go through 1 -> 2 -> 3. 

The summary is that we should support a direct 1 -> 3 upgrade path for the following reason:
Basically any upgrade of the format version depends on whether it is backwards compatible. Currently, a V2 table is fully backwards compatible with V1 since all V1 metadata is valid in V2. 
Following the same approach, V2 metadata should most likely be valid in V3 which implies that V1 metadata would be valid in V3 as well and thus a direct 1 -> 3 should be supported